### PR TITLE
feat(monitor): support monitor and alert

### PR DESCRIPTION
**Pull request by huangting4201**
## Motivation

Monitor training status, loss value, exception status and so on, send alert info to feishu.

## Modification

- `monitor/alert.py`: add func `send_feishu_msg_with_webhook` for sending alert info to feishu.
- `monitor/monitor.py`: add func `initialize_monitor` for initializing monitor thread.

## Use cases (Optional)

User can set feishu webhook alert address in `configs/7B_sft.py`:
```python
# If not set, monitor will be disabled
alert_address = "xxx"
```

Alert info example is shown as follows:
企业微信截图_4feb905d-b2f8-4ce3-b535-930e80da16d9

## Checklist

**Before PR**:

- [x] Pre-commit or other linting tools are used to fix the potential lint issues.

### CODE REVIEW

File: internlm/initialize/launch.py [hunk 1]
```diff
@@ -38,6 +38,11 @@ def get_default_parser():
     parser.add_argument("--local_rank", type=int, help="local rank on the node")
     parser.add_argument("--backend", type=str, default="nccl", help="backend for distributed communication")
     parser.add_argument("--seed", type=int, default=1024)
+    parser.add_argument("--job_name", default=None, type=str, help="The training job name.")
```
Comment (sunpengsdu) [inline internlm/initialize/launch.py:41]: 话说model config里有个JOB NAME了，两者怎么衡量

Comment (sunpengsdu) [reply internlm/initialize/launch.py:41]: monitor和alert_address是不是可以都统一放config里？

Comment (huangting4201) [reply internlm/initialize/launch.py:41]: 命令行传参新增job_name的原因是，发送告警信息时想用job_name的信息作为title，但是config文件中的信息，得main函数里初始化分布式环境后才能拿到，所以这个我就新加个这个参数

Comment (huangting4201) [reply internlm/initialize/launch.py:41]: 飞书地址作为敏感字段不想暴露，所以把alert_address也作为命令行参数传入，代码里也不设置默认值

File: internlm/monitor/monitor.py [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+def send_alert_message(address: str = FEISHU_WEBHOOK_ADDRESS, title: str = None, message: str = None):
+    if ENABLE_MONITOR:
+        send_feishu_msg_with_webhook(
+            webhook=address,
+            title=title if title else get_job_key(),
+            message=message,
+        )
+
+
+class MonitorTracker(Thread):
```
Comment (sunpengsdu) [inline internlm/monitor/monitor.py:39]:
```
from threading import Thread
import time
import os

class MonitorTracker(Thread):

    def __init__(self, webhook, loss_limit=2, check_interval=60):
        super().__init__()
        self.webhook = webhook
        self.loss_limit = loss_limit
        self.check_interval = check_interval
        
        self.last_loss = float('inf')
        self.last_active = 0
        self.stopped = False
        
    def run(self):
        while not self.stopped:
            self._check_stuck()
            self._check_loss_spike()
            time.sleep(self.check_interval)
            
    def _check_stuck(self):
        current_active = int(os.getenv('LAST_ACTIVE', 0))
        if current_active <= self.last_active:
            self._send_alert("Training may be stuck")
        self.last_active = current_active
            
    def _check_loss_spike(self):
        current_loss = float(os.getenv('CURRENT_LOSS', 0))
        if current_loss / self.last_loss > self.loss_limit:
            step = int(os.getenv('CURRENT_STEP', 0))
            self._send_alert(f"Loss spiked in step {step}")
        self.last_loss = current_loss
            
    def _send_alert(self, message):
        if self._is_master():
            send_message(self.webhook, message)
            
    def stop(self):
        self.stopped = True

```
这个类Claude处理后的样子，感觉可以借鉴下

Comment (huangting4201) [reply internlm/monitor/monitor.py:39]: 可以

File: internlm/monitor/monitor.py [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+    format_trace = ""
+    for line in filtered_trace:
+        format_trace += "\n" + line
+    send_alert_message(
+        address=FEISHU_WEBHOOK_ADDRESS,
+        message=f"Catch Exception from {socket.gethostname()} with rank id {gpc.get_global_rank()}:{format_trace}",
+    )
+
+
+def monitor_loss_spike(step_count, cur_step_loss):
```
[the remark this answers is not in the document]
Comment (huangting4201) [reply internlm/monitor/monitor.py]: 已更新

File: train.py [hunk 4]
```diff
@@ -646,9 +662,17 @@ def main(args):
...
+    # initialize distributed environment
+    initialize_distributed_env(config=args.config, launcher=args.launcher, master_port=args.port, seed=args.seed)
+    assert hasattr(gpc, "config") and gpc.config is not None
+
+    # initialize monitor manager context
+    with initialize_monitor_manager(job_name=gpc.config.JOB_NAME, alert_address=gpc.config.alert_address):
+        try:
+            main(args)
+        except Exception:
+            print(f"Raise exception from {hostname} with rank id: {gpc.get_global_rank()}")
```
Comment (sunpengsdu) [inline train.py:676]: print to log?

Comment (huangting4201) [reply train.py]: 已更新
